rfcs/45-stageless.md

```diff
@@ -0,0 +1,984 @@
+4. "Command-flushed" ordering constraints (dependencies that will ensure the pair is separated by command application).
+
+There are also lots of related but non-urgent areas for follow-up research:
+
+1. Improve tools for reporting and resolving system execution order ambiguities.
+2. Revisit plugin definition and plugin configuration.
+3. Cloning registered systems. (This probably warrants its own RFC, but it fits with the multiple worlds discussion.)
+4. Add [manual dependency graph construction](https://github.com/bevyengine/bevy/pull/2381) methods.
+5. Seeded, fully deterministic system execution orders: useful for debugging system order bugs and optimization work.
+6. Move command queues into the `World` so that exclusive systems can drain them directly (e.g. in some user-defined order).
+7. Compose conditions via arbitrary boolean expressions.
+8. Support automatic insertion and removal of systems to reduce schedule clutter and support other forms of one-off logic.
+9. Run schedules without `&mut World`, inferring access based on the systems inside.
+
+## Appendix
+
+These implementation ideas/details are here to avoid cluttering the RFC.
+
+### `apply_system_buffers`
+
+As command queues are owned by their systems, an exclusive system won't be able to drain their buffers by itself.
+This one just signals the executor to do it.
+The caveat to handling it this way is that the executor is only aware of the systems in the schedule it's running, so we should probably still have one "courtesy flush" after all the systems complete, so users can't forget to.
+
+### `Systems`
+
+`Systems` is a resource on the `World`.
+
+To keep `Systems` accessible, you have to take systems and schedules out of `Systems` in order to run them.
+Likewise, system insertion and removal will not immediately cause schedules to be rebuilt.
+Those changes will only be propagated up to flag the system sets whose dependency graphs need to be updated (topologically sorted again).
+A rebuild will remain pending until something tries to check out the `Schedule`.
+This way users only incur costs for schedules they use.
+
+Schedule extraction moves the required systems and conditions from their long-term storage (`Option`) into the selected `Schedule` container and returns it.
+
+Even after systems or schedules have been extracted, all the graph data remains in `Systems`, so you can still add and remove things.
+You still have to return extracted schedules to `Systems` in order to update them (since that process needs the systems), but if you had removed any of those systems, they will be safely dropped.
+
+```rust
+enum NodeId {
+    System(u64),
+    Set(u64),
+}
+
+#[derive(Resource)]
+pub struct Systems {
+    // `NodeId` is unique identifier, generated on insertion
+    index: HashMap<SystemLabelId, NodeId>,
+    next_id: u64,
+
+    // long-term storage
+    graphs: HashMap<NodeId, DependencyGraph>,
+    systems: HashMap<NodeId, Option<BoxedSystem>>,

```

Comment:
Hundreds of thousands of systems is an obscenely large number of systems. Possible, but unlikely. I'll retroactively scope that down to "thousands of systems".

Reply:
>  Is this done every frame or once per system as part of initial setup?

Based on my experiments with one-shot systems, I think this should be feasible to do during initial setup (or eventually, when schedules are modified). I'll defer to @maniwani for a firmer answer though.

Reply:
> How and when do systems get passed from `Systems` into `Schedule` instances? Is this done every frame or once per system as part of initial setup?

> Especially given that it seems like systems in the "main" schedule would need to be removed for the duration of the schedule, so the "dynamicness" of this seems pretty limited in practice.

I can't guess at the performance without benchmarks or profiling, but yes, as written, the systems would be shuffled back and forth between them. 

The "dynamicness" comes in (1) allowing schedules to be regenerated at runtime and (2) allowing any system or system set that isn't already removed (as you've correctly understood) to be retrieved and used on-demand (e.g. within a command).

> maybe keeping a clearer separation between system sets and schedules would allow us to compute things "once" and not need to shuffle potentially hundreds of thousands of systems around every frame

It doesn't make sense to separate sets and schedules (as they represent the same underlying concept), but we can probably just keep systems in their last-used schedules, as long as we track those locations if we do need to shuffle them around for some reason.

Reply:
> I don't think it makes sense to separate sets and schedules any further (as they represent the same underlying thing).

I think the fact that we need to convert between representations means that they aren't the same thing. System sets can be ordered relative to other system sets, but schedules (the collection of information required to run a group of systems) cannot be ordered relative to other schedules in the same way. Converting a system set to a schedule is a "lossy" conversion.

Can we go through the thought experiment of treating schedules as "buckets of systems/sets" and see if we can still cover all the scenarios we care about? It seems like it would work just as well for the things we need (OnEnter, fixed updates, main schedules, on demand system execution, etc), with a much clearer data model (both from a user facing perspective and an ownership perspective). All of these subtractive "set ops" that happen as part of running system sets as schedules are things users need to think about. They need to "do the math in their head" when it comes to system availability or they will hit runtime errors.  I don't think these set ops open any doors that should be opened. And they introduce a lot of weirdness and overhead.

Imo this data model is crisp and clear:

* System: a function that does work on World. can be ordered relative to other sets and systems. (a node with "order edges" and "set membership" edges)
* System Set: a set of systems that can be ordered and configured relative to other sets and systems (a node with "order edges" and "set membership" edges)
* Schedule: a runnable graph of systems and sets.
* Schedules: a name->schedule collection.

This data model is way less clear:

* System: a function that does work on World. can be ordered relative to other sets and systems. (a node with "order edges" and "set membership" edges)
* System Set: a set of systems that can be ordered and configured relative to other sets and systems (a node with "order edges" and "set membership" edges, but one that can be extracted from the graph into a new Schedule graph, temporarily invalidating the correctness of all related order edges and set membership edges, while still enabling users to try slicing the graph in other valid ways)
* Schedule: a temporarily runnable slice of the Systems graph generated on demand
* Systems: a graph of all systems and sets

Reply:
I think this pivots quite a bit from the topic of shuffling and performance. Is that no longer a concern? I think my suggestion to keep systems in their most recently used schedules is one valid solution for that. Do you agree?

On the question of data model, is this the essence of your proposal? You would rather have users essentially commit to specific schedules (i.e. only their root/top-level sets are runnable), rather than allow any system set to be "converted" into a schedule since that can fail?

I agree that simply having no flexibility here avoids a class of runtime errors, but I'm wondering if you're tossing useful functionality. With what you propose, could you get an individual system for those "one-shot" scenarios? Would you need to add them to 1-system schedules (that need their own names)? Like, I'm a bit skeptical if you're suggesting we should have both `Schedules` *and* `Systems`.

> Imo this data model is crisp and clear

I think the "crisp" model you've presented is identical to our data model, just with `Systems` swapped with `Schedules`.

This is how I see our model:

- system
  - a (named) function that does something with data in the world
  - a dependency graph node that can be ordered relative to other systems and sets
  - to run it, you extract it from `Systems` (temporarily)
- system set
  - a named group of nodes (systems and other sets)
  - a dependency graph "super" node that can be ordered relative to other systems and sets
- schedule
  - a runnable version of a system set
  - to run it, you extract it from `Systems` (temporarily, grabs all the systems needed)
- systems
  - stores all system-related stuff
  
Both systems and system sets have graph-related data and executable elements. Likewise, in both cases, the executable element is the thing you extract. For systems, you get a `Box<dyn System>`, for sets you get a `Schedule`. Therefore, in this model, "ordering schedules" has no meaning. And I don't think "conversion" is the right word. You're just extracting the system set's "function".

I'm thinking it may have been clearer to not have two names for it. Seems like it's masking the symmetry.

> generated on demand

I consider this an implementation detail rather than "public knowledge". We *could* just generate all `Schedule` instances upfront and there'd be no observable difference outside of profiling. We can choose to save work or not.

> [extracting a schedule] temporarily invalidates the correctness of all related order edges and set membership edges

This seems incorrect. Shuffling systems into a schedule does not "invalidate" the dependency graph. The only effect is that those systems become unavailable. And I don't think the "mental math" around that is all complicated either.

If a schedule is running, those systems are just unavailable. I would think sets should *help* here. Hopefully users rarely have to mentally account for lots of individual systems. We have the ability to provide very clear error messages here (i.e. we can tell you where your systems went).

Reply:
> I think this pivots quite a bit from the topic of shuffling and performance. Is that no longer a concern?

I've pivoted because I think the more important issue to resolve is the data model. The performance concerns stem from that, but they are one (potentially treatable) symptom of what I perceive to be a larger problem. 

>  I think my suggestion to keep systems in their most recently used schedules is one valid solution for that. Do you agree?

This would probably help in the case where "system sets run as Schedules" are totally disjoint within an app, but the second you try actually using the extra "overlapping set + schedule" flexibility unlocked by system, the concerns start to resurface. We can do our best to optimize this, but the optimal performance of the model is still bounded by the size of the set intersections and the number of intersecting schedules run. This is concerning because the plan involves hiding this set difference complexity from the user and making it _very easy_ to do "set difference" ops.

> On the question of data model, is this the essence of your proposal? You would rather have users essentially commit to specific runnable graphs (i.e. only the root/top-level sets would be runnable), rather than allow any system set to be "converted" into a schedule, since that can fail?

Yup that is what I am proposing.

> I agree that not having this "slicing" flexibility avoids a class of runtime errors, but idk if we won't want it later.

If we can come up with scenario that is valuable enough to offset what I perceive to be a high (app dependent) perf cost and the risk of "cannot run this system set as system X is currently claimed" logical errors, I would be more convinced.

> This should still be possible as well. I don't think we should store the dependency graph(s) in the runnable Schedule struct and let you mutate it directly though. You'd be asking users to remember where everything is again.

I'm pretty certain that this is possible, given that our current schedule impl supports this. I rather like the current SystemStage approach: you can add new systems at any time (provided you have a mutable reference), which invalidates the internal optimized dependency graph representation / recomputes as necessary.  Of course, getting the mutable reference to the schedule from within App execution is currently impossible on main. But thats solved by the Systems (or Schedules) resource proposed here. 

Seems like that approach generalizes nicely to stageless (ex: rename SystemStage to Schedule, adapt it to do stageless logic). Not actually proposing that we take that implementation path, just making a point.

> This [running systems on demand] is a little unclear [in the "schedules are completely disjoint and separate from system sets model"]. I don't think we should have a Schedules and a Systems, so with what you propose, can you still extract individual systems for those "one-shot" usecases or would we special-case them as one-element schedules?

Yeah I agree that having both Schedules and Systems doesn't make sense. I showed that we could have a rough equivalent to the RFC's proposed API in [this comment](https://github.com/bevyengine/rfcs/pull/45#discussion_r953198893). You just replace the "set label" for the group of systems with a "schedule label".

That doesn't involve true "on-shot systems" (running individual systems addressed by their names instead of their sets), but I don't see an equivalent in this RFC either. I would argue that actual "one shot" systems are already a solved problem: just use `System::run` in whatever context owns the system. `world.schedule_scope(some_system, |world, schedule| { schedule.run(world) })` is way more complicated than that.  

> I think the model you've presented is nearly identical to ours, just with Systems swapped with Schedules.

I agree, as long as we also agree that "dynamically generated Schedules from system sets subtracted from a larger dependency graph" is not a trivial design difference.

> Both systems and system sets have graph-related info and an executable element. In both cases, the executable element is what you extract. For systems, you get a Box<dyn System>, for sets you get a Schedule. Therefore, in this model, "ordering schedules" has no meaning. I also don't think "conversion" is the right word. You're just extracting the system set's "function". Maybe it would have been better to scrap "schedule" and only say "system sets".

Consider the following App to see why I think "ignoring" dependencies for a SystemSet extracted into a Schedule is wrong (or at the very least, an action that drops relevant information):

```rust
app 
  .add_set(ProduceOutputs)
  .add_set(ConsumeOutputs)
  .add_systems(systems![create_foo, create_bar].in_set(ProduceOutputs))
  .add_systems(systems![consume_foo, consume_bar].in_set(ConsumeOutputs).after(ProduceOutputs))
  .add_system(manual_run);

fn manual_run(world: &mut World) {
 // this runs ConsumeOutputs, despite the fact that ProduceOutputs didn't run. we're violating the constraint
 // defined in our App schedule!
  world.schedule_scope(ConsumeOutputs, |world, mut schedule| {
    schedule.run(world);
  });
}
```

Now _maybe_ `ConsumeOutputs` was designed in a way to properly accommodate this scenario. But that is a "new' constraint that exists outside of the App as it was defined. And I would argue for pretty much every SystemSet, they are designed to work only within the bounds of the constraints as defined in the schedule. And most importantly, we are actively making it easy for _consumers_ of plugins to violate the constraints configured, without any warning at all.

By making Schedules "separate named graphs of sets and systems to be run", we are creating a contract that can be honored by consumers. I'm asserting that any slice of the global systems/set graph that violates a dependency edge is a violation of a contract, which we should not embrace as a supported pattern.

> I consider this an implementation detail rather than "general user knowledge". We could just generate all Schedule instances upfront and there'd be no observable difference outside of tracing. We can choose to save work or not.

I agree that this can be entirely hidden from users, but for apis with performance implications, thats not always a good thing. With the current ownership model, we'd still need to pass around systems / do a lot of copies / moves / identity hashes (unless you're arguing that systems should be stateless / each schedule should have their own copy, which I don't think is the move). At this point, while I think the perf implications of this design are concerning (even with ideal caching in place), I'd prefer to table the conversation until we've resolved the "high level data model" conversation, as that feels easier to hash out without going deep into technical specifics. Happy to call this argument "a draw" until then.

> Similar story here too. I wouldn't say Systems is a graph, but it does maintain "one" internally so you can add and remove things after the app has started (and ig for visualizations as well).

`Systems` owns the data that is the source of truth for dependency information for the entire app. I would personally consider it to be a graph, but thats certainly a matter of perspective. I'm happy to call it something like "the global owning container of all system and set information, including dependencies between them" if thats preferable.

>  This seems incorrect. Shuffling systems into a schedule does not "invalidate" the dependency graph. The only effect is that those systems become unavailable.

I disagree. See the example I provided above that illustrates what I consider to be "invalidation".

> And I don't think the mental math there is all that complicated either

This depends on the situation. We're allowing arbitrary nesting of "system subtraction" logic, potentially across multiple parties. If you are nested in a context that another plugin defined, how are you supposed to know what systems are still available within the global context? What if they push an update that claims a new system, breaking your code? And even if you own all of the logic, once you're a couple of levels of "sets consumed" deep, it could be hard to construct the set of "currently available" systems in your head. This is a can of worms that I am very hesitant to open.

Reply:
> Consider the following `App` to see why I think "ignoring" dependencies for a `SystemSet` extracted into a `Schedule` is wrong (or at the very least, an action that drops relevant information):

I think I've come to agree that slicing is bad. That said, I think you misunderstood something somewhere because your example would panic.

If `ConsumeOutputs` and `manual_run` are both part of the app update, the extraction inside `manual_run` is going to fail because the `ConsumeOutput` systems will have already been extracted.

Also, this part was a bit non-idiomatic.

```rust
.add_systems(
    systems![consume_foo, consume_bar]
        .in_set(ConsumeOutputs)
        .after(ProduceOutputs)
)
```

This use of `after` would hit case (5) in the graph checker (if set to warn or error). It happens to be valid here, but you should really define this dependency on `ConsumeOutputs` or `ProduceOutputs`.

> Seems like that approach [of adding systems directly to the runnable container] generalizes nicely to stageless (ex: rename `SystemStage` to `Schedule`, adapt it to do stageless logic). Not actually proposing that we take that implementation path, just making a point.

Yeah, I wasn't talking about whether *this* was possible, but I don't want this implementation path. I don't like the idea of the dependency graphs moving around.

Earlier, you said:

> All of these subtractive "set ops" that happen as part of running system sets as schedules are things users need to think about. They need to "do the math in their head" when it comes to system availability or they will hit runtime errors.

`Schedule` working like `SystemStage` would have the same issue.

Maybe not as much, but users would still need to "do the math in their head or hit runtime errors" if they need to know which schedule to modify.

I understand that it's a neat parallel, but do you think schedules owning their dependency graphs and being directly malleable is actually more useful than `Schedules` owning them?

Because I would argue that it's simpler if graph modifications always go through `Schedules`.

- `App::add_system` and `App::add_set` methods would already go through `Schedules`.
- Maybe you'll say this isn't an ability we should have, but if you want to queue up modifications to a schedule that's running or otherwise unavailable, your only option is to go through `Schedules`. Seems a bit clunky to have two places to add systems.
- You'll encounter fewer runtime errors from forgetfulness if changes go through `Schedules` since it'll know which schedule to update. (Or worse, silent failures since I think you also want that "missing sets default to empty" behavior.)

> If we can come up with scenario that is valuable enough to offset what I perceive to be a high (app dependent) perf cost and the risk of "cannot run this system set as system X is currently claimed" logical errors, I would be more convinced.

I can't think of anything in particular. I think any case where you'd want to "slice" could probably be addressed by just adding everything again under a different schedule.

> I agree, as long as we also agree that "dynamically generated Schedules from system sets subtracted from a larger dependency graph" is not a trivial design difference.

Sure, the mental model is appreciably different, but implementation-wise, we can bridge from the current proposal.

Just restrict schedule extraction to the root sets. Maybe you'd have to call a `root` method or something anyway if adding things has a default like `CoreSystems::Update`. Likewise, system extraction can be restricted to systems that were added standalone. 

That seems equivalent and presumably would not involve any shuffling beyond the initial construction. I think the only major difference then is what goes in the `Schedule` struct.

> And I would argue for pretty much every `SystemSet`, they are designed to work only within the bounds of the constraints as defined in the schedule. And most importantly, we are actively making it easy for consumers of plugins to violate the constraints configured, without any warning at all.

> By making Schedules "separate named graphs of sets and systems to be run", we are creating a contract that can be honored by consumers. I'm asserting that any slice of the global systems/set graph that violates a dependency edge is a violation of a contract, which we should not embrace as a supported pattern.

Yeah, I think I've come around. I agree in principle that slicing isn't worth the increase in runtime errors.

> That doesn't involve true "on-shot systems" (running individual systems addressed by their names instead of their sets), but I don't see an equivalent in this RFC either.

Ah, we just forgot to write in a system version of `schedule_scope`, but that's trivial to add and would work the same way. Pass in a label and a function that handles a `&mut Box<dyn System>`.

Simply taking the systems and schedules (and having to manually return them) is also possible.

> I would argue that actual "one shot" systems are already a solved problem: just use `System::run` in whatever context owns the system.

I disagree. I think you're hand-waving the need to store those systems somewhere. How does a context get ownership of a system?

If you're saying "just create and initialize the systems there" in the context (e.g. in a command), that's not a solution. Systems have persistent state (locals, change ticks) that users want to make use of.

Reply:
> Systems have persistent state (locals, change ticks) that users want to make use of.

This was one of the critical things I discovered in my one-shot systems work. We ultimately need to store systems *somewhere* in order to avoid very surprising edge cases.

Reply:
> If ConsumeOutputs and manual_run are both part of the app update, the extraction inside manual_run is going to fail because the ConsumeOutputs systems will have already been extracted.

This thought crossed my mind after first writing it (I caught this bug on a first pass over my post), but I chose not to modify it or address it because I trusted that it still made the point I was trying to make about constraints / that given the depth of the conversation we're having about this specific issue, that I would be given the benefit of the doubt. From your response I think it accomplished the intended goal! I do understand how this works and I'm not confused (but I do appreciate that you followed up / checked).

Sounds like I've converted you to the labeled-schedule-driven execution model! @alice-i-cecile are you also on board?

> Schedule working like SystemStage would have the same issue.

The point is that we agree that SystemSets should not be the unit of execution, as that encourages graph slicing. I agree that subtracting a full "unit of execution" Schedule from Schedules suffers from the same class of problem. But the difference is that we aren't blurring the lines between "groups of systems ordered relative to other groups of systems" and "units of execution". A "unit of execution" is something defined to be claimed and run at some point in the app. The buck needs to stop somewhere.

It sounds like we agree on that now, so I'm not sure what 

>  I understand that it's a neat parallel, but do you think schedules owning their dependency graphs and being directly malleable is actually more useful than Schedules owning them?

It isn't just a neat parallel. It accurately models the hard lines between schedules (preventing accidental blurring of those lines) and it removes the (completely unnecessary) need for passing unbounded amounts of data around on each run.

> Maybe you'll say this isn't an ability we should have, but if you want to queue up modifications to a schedule that's running or otherwise unavailable, your only option is to go through Schedules. Seems a bit clunky to have two places to add systems.

I'm totally on board for queuing up changes to a schedule at arbitrary points. But the authoritative representation of schedule that is currently running shouldn't be directly modified (which includes the systems running in it). I think queuing up systems to be added _should_ be a different api than adding a system directly to a schedule. And we should support both operations. I see no harm in making `Schedules.queue_add_system(foo.in_schedule(SomeSchedule))` explicit. That feels like a win.

Making "system add ops" happen directly on a schedule that you have direct access to (and is the authority on the data model) means that you can do more data validation / compute optimal data representations on the spot. And it removes the complexity of "deferred data integrity" and "moving data around".

> Sure, the mental model is appreciably different, but implementation-wise, we can bridge from the current proposal. Just restrict schedule extraction to the root sets. 

I personally think a new, separate concept (the Schedule ... which owns the systems, sets, and associated constraints it runs) is preferable to trying to "promote" some sets as special / roots. From my perspective, system sets are _not_ graphs. They are nodes in a graph (which can have edges). If we "promote" some sets, then to avoid slicing we need to prevent ordering constraints between them, which will require another set of _runtime enforced_ validation rules. Why go through all of that trouble (and user facing confusion) when we can get the same protections at compile time by using a more accurate data model?

I think "generating" schedules at runtime isn't worth the cost (both performance-wise and data model wise).  From my perspective, we should re-build this part in the current implementation.

> Ah, we just forgot to write in a system version of schedule_scope, but that's trivial to add and would work the same way. Pass in a label and a function that handles a &mut Box<dyn System>.

Makes sense!

> If you're saying "just create and initialize the systems there" in the context (e.g. in a command), that's not a solution. Systems have persistent state (locals, change ticks) that users want to make use of.

The case I was making there is that the "context" would already own the system. For example, we could use "closure systems" for this:

```rust
app.add_system(bar());

fn foo(query: Query<&Transform>) { }

fn bar() -> impl FnMut(&mut World) {
  let foo = System::into_system(foo);
  move |world| {
     if world.resource::<Time>().seconds_since_startup > 2.0 {
       foo.run(world);
     }
  }
}
``` 

Or we could use Resources for this:

```rust
#[derive(DerefMut)]
struct Foo(Box<dyn System>);

impl Default for Foo {
  fn default() -> Self {
    Self(System::into_system(foo))
  }
}

fn bar(world: &mut World) {
  world.resource_scope(|world, mut foo: Mut<Foo>| {
    foo.run(world);
  });
}
```

Or we could _try_ doing something like a `LocalSystem<foo>` SystemParam, but im pretty sure that relies on unstable / unimplement const features and/or would require more type parameters.

That being said, the "one system in a schedule" pattern also seems perfectly fine.

```rust
#[derive(ScheduleLabel)]
struct Foo;

app.add_system(foo.in_schedule(Foo));

fn bar(world: &mut World) {
  // seems like we would want this api as a shorthand for schedule_scope anyway
  world.run_schedule(Foo);
}
```

Notably, this requires zero refactoring if it is later decided that doing "foo" requires more than one system. Just add another system to Foo.

Reply:
> I still think having the unit of execution and the graph representation lumped together is a little weird but *shrug*.

You might be right about this. I currently think treating this as one data structure (which owns the systems) makes the most sense, as it works pretty well for the current SystemStage impl,  it lets us do as much work up front as possible, and it avoids the need to "sync" some things up / pass data around across structures. Looking at a "separated impl" might change my mind (after accounting for the changes we've already agreed on), but with my current understanding of the problem I think the "one data structure model" is the move. Happy to defer this choice to the implementation PR, but im pretty biased in one direction atm.

> Idk how else I could have responded. You couldn't double extract. It's hard to read between the lines of an invalid example when there's so much nuance in what's being said.

Totally fair. I also think I phrased that in a way that was unfair to you / implied that you weren't operating in good faith. You definitely were/are. Sorry about that!

> I just want you to recognize that this is a novel interpretation. It doesn't just follow naturally that dependencies are relevant when something is called outside the graph. (Maybe we disagree on what "outside the graph" means.) But if we take that as truth, then sure, it's one more reason against being able to extract any arbitrary system set.

I recognize that this is one of many valid interpretations (and I cede that the original interpretation of this RFC _is_ valid), but from my perspective in the context of dependencies between things, treating those dependencies as optional based on the context is the more novel interpretation. Like if you look at a Rust program's Cargo dependencies, it feels safe to assume that if you don't compile those dependencies first, bad things might happen. Obviously not fully apples to apples, but I think this is the default way that people think about dependencies. _And_ its the way people will naturally use them in the context of Bevy (ex: System A depends on the results from System B, so running System A without running System B first will likely break something. The dependency exists for a functional reason.)

> I think this wasn't an issue either way? If schedules worked as described in the current revision, the graphs are stored separately and are safe to modify. If schedules own their graphs, then you can't mutably access them while running.

Agreed. The differences are reasonably superficial. (1) whether or not we apply a change immediately to the source of truth or queue it up for later / what we define the source of truth to be (2) Whether or not we expose both "immediate" and "queued" apis (3) what we call those apis.

> Either way, once you start allowing changes at runtime, then whenever users make them, we will have to (re)allocate space and topsort and check graphs for errors (again).

Yup that makes sense. I do agree that there will probably always be some schedule generation work that need to be "queued". I just also think that we'll benefit from doing that queued work "internally" / allowing the graph generation / executor logic to reuse work and storage from the user-defined Schedule type.